ghcide/src/Development/IDE/LSP/LanguageServer.hs

```diff
@@ -57,11 +58,16 @@ runLanguageServer
 runLanguageServer options inH outH getHieDbLoc defaultConfig onConfigurationChange userHandlers getIdeState = do
 
     -- This MVar becomes full when the server thread exits or we receive exit message from client.
-    -- LSP loop will be canceled when it's full.
+    -- LSP server will be canceled when it's full.
     clientMsgVar <- newEmptyMVar
     -- Forcefully exit
     let exit = void $ tryPutMVar clientMsgVar ()
 
+    -- An MVar to control the lifetime of the reactor loop.
+    -- The loop will be stopped and resources freed when it's full
+    reactorLifetime <- newEmptyMVar
+    let stopReactorLoop = void $ tryPutMVar reactorLifetime ()
```

Comment:
Why `tryPutMVar`? So it doesn't block? Shouldn't we at least return the bool and log something in the shutdown handler if it fails to shutdown the reactor?

Reply:
Why? So that it can be called twice (not that we depend on that)
Not sure which bool you have in mind.
The reactor is guaranteed to be killed, and it has an exception handler attached (which hopefully should only capture non-async exceptions)

Reply:
We already log a "Received shutdown message", that's good enough for me